test/unit/tcti-device.c

```diff
@@ -63,6 +79,63 @@ tcti_device_init_size_test (void **state)
     ret = Tss2_Tcti_Device_Init (NULL, &tcti_size, NULL);
     assert_int_equal (ret, TSS2_RC_SUCCESS);
 }
+/* Test the failure of opening a specified device file */
+static void
+tcti_device_init_conf_fail (void **state)
+{
+    size_t tcti_size = 0;
+    TSS2_RC ret = TSS2_RC_SUCCESS;
+    TSS2_TCTI_CONTEXT *ctx = NULL;
+
+    ret = Tss2_Tcti_Device_Init (NULL, &tcti_size, NULL);
+    assert_true (ret == TSS2_RC_SUCCESS);
+    ctx = calloc (1, tcti_size);
+    assert_non_null (ctx);
+    errno = ENOENT; /* No such file or directory */
+    will_return (__wrap_open, -1);
+    ret = Tss2_Tcti_Device_Init (ctx, &tcti_size, "/dev/nonexistent");
+    assert_true (ret == TSS2_TCTI_RC_IO_ERROR);
+
+    free(ctx);
+}
+/* Test the device file recognition if no config string was specified */
+static void
+tcti_device_init_conf_default_fail (void **state)
+{
+    size_t tcti_size = 0;
+    TSS2_RC ret = TSS2_RC_SUCCESS;
+    TSS2_TCTI_CONTEXT *ctx = NULL;
+
+    ret = Tss2_Tcti_Device_Init (NULL, &tcti_size, NULL);
+    assert_true (ret == TSS2_RC_SUCCESS);
+    ctx = calloc (1, tcti_size);
+    assert_non_null (ctx);
+    errno = EACCES; /* Permission denied */
+    will_return (__wrap_open, -1);
```

Comment:
Looks like one will_return is redundant here.

Reply:
This is needed here. Passing a `NULL` conf will result in trying to open the specified default files one after another until `open()` succeeds. See the [call to open()](https://github.com/tpm2-software/tpm2-tss/blob/051160c39db6215a8931887ce677cc2040e9b035/src/tss2-tcti/tcti-device.c#L459) inside of the for loop.

These default files are:

- `/dev/tpm0`
- `/dev/tpmrm0`

If both attempts fail (thus returning `-1` twice) an `TSS2_TCTI_RC_IO_ERROR` is returned by `Tss2_Tcti_Device_Init()`.